more e2eservice.TestJig cleanups

danwinship opened the pull request:
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
`e2eservice.TestJig` has a method `SanityCheckService` which is called in many places after creating or updating a service, and there's really no good reason to not just call it automatically from the end of the create/update methods instead.

(EDIT: additional cleanups were added beyond just this)

```release-note
NONE
```

/priority backlog

danwinship committed:
Remove an unneccessary e2eservice.TestJig method

danwinship committed:
Add Namespace to e2eservice.TestJig, make all methods use jig namespace and name

Many TestJig methods made the caller pass a serviceName argument, even
though the jig already has a name, and every caller was passing the
same name to each function as they had passed to NewTestJig().

Likewise, many methods made the caller pass a namespace argument, but
only a single test used more than one namespace, and it can easily be
rewritten to use two test jigs as well.

danwinship committed:
Call jig.SanityCheckService automatically after changes

danwinship commented:
/retest

timothysc commented:
/assign @alejandrox1

danwinship commented:
/retest

alejandrox1 reviewed (changes requested):
This looks good! Thank you @danwinship for working on this.
Left some comments.

danwinship commented on test/e2e/framework/service/jig.go:
```diff
@@ -161,6 +163,7 @@ func (j *TestJig) CreateExternalNameServiceOrFail(namespace string, tweak func(s
 	if err != nil {
 		framework.Failf("Failed to create ExternalName Service %q: %v", svc.Name, err)
 	}
+	j.sanityCheckService(result, v1.ServiceTypeExternalName)
```
In theory, the argument is being used to double-check that the service was actually created as the type we asked it to be created as, or in other places, that updating the service didn't cause its type to be changed. Yes, it's a bit superfluous, but most of `sanityCheckService` is "superfluous" in the same way... Anyway, I'll fix this one.

danwinship commented on test/e2e/framework/service/jig.go:
```diff
@@ -175,7 +178,11 @@ func (j *TestJig) CreateServiceWithServicePort(labels map[string]string, namespa
+	if err == nil {
+		j.sanityCheckService(result, service.Spec.Type)
+	}
+	return result, err
```
Hah. In my original patch I removed that function because it was so mismatched, but then I undid that change to keep the PR simple. I'll add it back...

danwinship commented on test/e2e/framework/service/jig.go:
```diff
@@ -378,12 +384,12 @@ func (j *TestJig) SanityCheckService(svc *v1.Service, svcType v1.ServiceType) {
 		if svc.Spec.ExternalName != "" {
 			framework.Failf("unexpected Spec.ExternalName (%s) for service, expected empty", svc.Spec.ExternalName)
 		}
-		if svc.Spec.ClusterIP != api.ClusterIPNone && svc.Spec.ClusterIP == "" {
```
It's redundant; if `ClusterIP == ""` then necessarily `ClusterIP != None`.

(The original author presumably wrote it that way because they were thinking "if ClusterIP isn't None, then make sure the service got assigned a ClusterIP".)

alejandrox1 commented on test/e2e/framework/service/jig.go:
```diff
@@ -161,23 +163,10 @@ func (j *TestJig) CreateExternalNameServiceOrFail(namespace string, tweak func(s
-		},
-	}
-	return j.Client.CoreV1().Services(namespace).Create(service)
-}
```
This we have to keep, otherwise it takes out an essential step from the e2e tests where it is used.
And looking at the e2e tests without this brings out a good point.

If a method creates something, it is great if that method checks that the created object is correct and what we would expect but the sanityChecks fail in the presence of an error.
There is also the fact that `test/e2e/framework/service` is supposed to be a library for users to consume, not "test code" so failing is a bit weird.

I think we should actually rewrite `sanityCheckService()` to return an error, instead of calling any `framework.Fail()`.
The rewrite all the jig create* methods to call `anityCheckService()` within them and return that error for the end-user to handle.

So for example, rewrite the create methods to something like this:
```go
func (j *TestJig) CreateServiceWithServicePort(labels map[string]string, namespace string, ports []v1.ServicePort) (*v1.Service, error) {
	service := &v1.Service{
		ObjectMeta: metav1.ObjectMeta{
			Name: j.Name,
		},
		Spec: v1.ServiceSpec{
			Selector: labels,
			Ports:    ports,
		},
	}
	result, err j.Client.CoreV1().Services(namespace).Create(service)
        if err != nil {
		return result, fmt.Errorsf("Failed to create TCP Service %q: %v", svc.Name, err)
	}
	err = j.sanityCheckService(result, svc.Spec.Type)
        return result, err
}
```

alejandrox1 commented on test/e2e/framework/service/jig.go:
```diff
@@ -378,12 +366,12 @@ func (j *TestJig) SanityCheckService(svc *v1.Service, svcType v1.ServiceType) {
 	} else {
 		if svc.Spec.ClusterIP != "" {
-			framework.Failf("unexpected Spec.ClusterIP (%s) for ExternamName service, expected empty", svc.Spec.ClusterIP)
+			framework.Failf("unexpected Spec.ClusterIP (%s) for ExternalName service, expected empty", svc.Spec.ClusterIP)
```
Going along with my other comment, if we are going to include within the jig create* methods (which would be good), we should rewrite this to return errors instead of making a decision for the user.
i.e.,
```go
if svc.Spec.ClusterIP != "" {
	return fmt.Errorsf("unexpected Spec.ClusterIP (%s) for ExternamName service, expected empty", svc.Spec.ClusterIP)
}
```

alejandrox1 reviewed (commented):
This latest patch of changes really helps add context to this. 
I think that besides running the sanity checks every time something is created, we should also modified the api for this methods to return the created objectd AND an error for the user to handle.

danwinship commented:
>  I think that besides running the sanity checks every time something is created, we should also modified the api for this methods to return the created objectd AND an error for the user to handle.

Yes, I was planning on doing that, but in a separate PR... I was trying to keep them small-ish and more-easily-reviewable. Do you want me to add in return-an-error-rather-than-failing patches to this PR now or would it be better as a separate PR?

danwinship committed:
Make e2eservice.TestJig methods return errors rather than failing

This is the preferred style for framework methods, and it lets us
write tests to confirm that certain things *don't* work.

alejandrox1 commented:
> > I think that besides running the sanity checks every time something is created, we should also modified the api for this methods to return the created objectd AND an error for the user to handle.
> 
> Yes, I was planning on doing that, but in a separate PR... I was trying to keep them small-ish and more-easily-reviewable. Do you want me to add in return-an-error-rather-than-failing patches to this PR now or would it be better as a separate PR?

Let's do it in this PR.
I think it makes more sense putting it all together.

danwinship commented:
/retest
@alejandrox1 pushed two new commits, one to return errors instead of failing, and then a second to remove redundant namespace and service name arguments from the jig functions

alejandrox1 commented on test/e2e/network/firewall.go:
```diff
@@ -82,15 +82,17 @@ var _ = SIGDescribe("Firewall rule", func() {
 			svc.Spec.Ports = []v1.ServicePort{{Protocol: v1.ProtocolTCP, Port: firewallTestHTTPPort}}
 			svc.Spec.LoadBalancerSourceRanges = firewallTestSourceRanges
 		})
+		framework.ExpectNoError(err)
```
could you add a simple annotation here as well, please?

alejandrox1 commented on test/e2e/network/firewall.go:
```diff
@@ -82,15 +82,17 @@ var _ = SIGDescribe("Firewall rule", func() {
 				svc.Spec.Type = v1.ServiceTypeNodePort
 				svc.Spec.LoadBalancerSourceRanges = nil
 			})
+			framework.ExpectNoError(err)
```
could you add a simple annotation here as well, please?

alejandrox1 commented on test/e2e/network/firewall.go:
```diff
@@ -116,9 +118,10 @@ var _ = SIGDescribe("Firewall rule", func() {
+		svc, err = jig.UpdateService(func(svc *v1.Service) {
 			svc.Spec.ExternalTrafficPolicy = v1.ServiceExternalTrafficPolicyTypeLocal
 		})
+		framework.ExpectNoError(err)
```
could you add a simple annotation here as well, please?

alejandrox1 commented on test/e2e/network/network_tiers.go:
```diff
@@ -65,17 +65,19 @@ var _ = SIGDescribe("Services [Feature:GCEAlphaFeature][Slow]", func() {
 		ginkgo.By("creating a pod to be part of the service " + svcName)
-		jig.RunOrFail(ns, nil)
+		_, err := jig.Run(nil)
+		framework.ExpectNoError(err)
```
could you add a simple annotation here as well, please?

alejandrox1 commented on test/e2e/network/network_tiers.go:
```diff
@@ -65,17 +65,19 @@ var _ = SIGDescribe("Services [Feature:GCEAlphaFeature][Slow]", func() {
 			svc.Spec.Type = v1.ServiceTypeLoadBalancer
 			setNetworkTier(svc, string(gcecloud.NetworkTierAnnotationStandard))
 		})
+		framework.ExpectNoError(err)
```
could you add a simple annotation here as well, please?

alejandrox1 commented on test/e2e/network/network_tiers.go:
```diff
@@ -84,21 +86,22 @@ var _ = SIGDescribe("Services [Feature:GCEAlphaFeature][Slow]", func() {
+		svc, err = jig.UpdateService(func(svc *v1.Service) {
 			clearNetworkTier(svc)
 		})
+		framework.ExpectNoError(err)
```
could you add a simple annotation here as well, please?

alejandrox1 commented on test/e2e/network/network_tiers.go:
```diff
@@ -119,31 +122,27 @@ var _ = SIGDescribe("Services [Feature:GCEAlphaFeature][Slow]", func() {
 			svc.Spec.LoadBalancerIP = requestedIP
 			setNetworkTier(svc, string(gcecloud.NetworkTierAnnotationStandard))
 		})
+		framework.ExpectNoError(err)
```
could you add a simple annotation here as well, please?

alejandrox1 commented on test/e2e/network/network_tiers.go:
```diff
@@ -119,31 +122,27 @@ var _ = SIGDescribe("Services [Feature:GCEAlphaFeature][Slow]", func() {
+	// If existingIP is "" this will wait for any ingress IP to show up. Otherwise
+	// it will wait for the ingress IP to change to something different.
+	svc, err := jig.WaitForNewIngressIP(existingIP, waitTimeout)
+	framework.ExpectNoError(err)
```
could you add a simple annotation here as well, please?

alejandrox1 commented on test/e2e/network/service.go:
```diff
@@ -124,24 +124,15 @@ var _ = SIGDescribe("Services", func() {
-
-		framework.ExpectNoError(err, "failed to create service with ServicePorts in namespace: %s", ns)
+		_, err := jig.CreateTCPServiceWithPort(nil, 80)
+		framework.ExpectNoError(err)
```
could you add a simple annotation here as well, please?

alejandrox1 commented on test/e2e/network/service.go:
```diff
@@ -124,24 +124,15 @@ var _ = SIGDescribe("Services", func() {
-		_, err := jig.CreateServiceWithServicePort(labels, ns, ports)
-
-		framework.ExpectNoError(err, "failed to create service with ServicePorts in namespace: %s", ns)
+		_, err := jig.CreateTCPServiceWithPort(nil, 80)
```
Why delete `CreateServiceWithServicePort()`?
Also, it looks like we are not building the service with the labels `foo=bar` and `baz=blah`.

alejandrox1 commented on test/e2e/network/service.go:
```diff
@@ -187,33 +178,33 @@ var _ = SIGDescribe("Services", func() {
+				},
+			}
+		})
+		framework.ExpectNoError(err)
```
could you add a simple annotation here as well, please?

alejandrox1 commented on test/e2e/network/service.go:
```diff
@@ -187,33 +178,33 @@ var _ = SIGDescribe("Services", func() {
-		}
-		_, err := jig.CreateServiceWithServicePort(labels, ns, ports)
-		framework.ExpectNoError(err, "failed to create service with ServicePorts in namespace: %s", ns)
+		_, err := jig.CreateTCPService(func(service *v1.Service) {
```
How are we specifying the labels here (`foo=bar`)?
it looks like they are not being set.

alejandrox1 commented on test/e2e/network/service.go:
```diff
@@ -282,10 +273,10 @@ var _ = SIGDescribe("Services", func() {
-		tcpService := jig.CreateTCPServiceWithPort(ns, nil, int32(servicePort))
-		jig.SanityCheckService(tcpService, v1.ServiceTypeClusterIP)
+		tcpService, err := jig.CreateTCPServiceWithPort(nil, int32(servicePort))
+		framework.ExpectNoError(err)
```
could you add a simple annotation here as well, please?

alejandrox1 commented on test/e2e/network/service.go:
```diff
@@ -353,10 +344,10 @@ var _ = SIGDescribe("Services", func() {
-		svc := jig.CreateTCPServiceWithPort(ns, nil, int32(servicePort))
-		jig.SanityCheckService(svc, v1.ServiceTypeClusterIP)
+		svc, err := jig.CreateTCPServiceWithPort(nil, int32(servicePort))
+		framework.ExpectNoError(err)
```
could you add a simple annotation here as well, please?

alejandrox1 commented on test/e2e/network/service.go:
```diff
@@ -372,7 +363,8 @@ var _ = SIGDescribe("Services", func() {
 		ginkgo.By("Checking if the pod can reach itself")
-		jig.CheckServiceReachability(ns, svc, pod)
+		err = jig.CheckServiceReachability(svc, pod)
+		framework.ExpectNoError(err)
```
could you add a simple annotation here as well, please?

alejandrox1 commented on test/e2e/network/service.go:
```diff
@@ -541,18 +533,21 @@ var _ = SIGDescribe("Services", func() {
-		jig.CreateServicePods(cs, ns, 2)
+		framework.ExpectNoError(err)
+		err = jig.CreateServicePods(2)
+		framework.ExpectNoError(err)
```
could you add a simple annotation here as well, please?

alejandrox1 commented on test/e2e/network/service.go:
```diff
@@ -541,18 +533,21 @@ var _ = SIGDescribe("Services", func() {
 		execPod := e2epod.CreateExecPodOrFail(cs, ns, "execpod", nil)
-		jig.CheckServiceReachability(ns, nodePortService, execPod)
+		err = jig.CheckServiceReachability(nodePortService, execPod)
+		framework.ExpectNoError(err)
```
could you add a simple annotation here as well, please?

alejandrox1 commented on test/e2e/network/service.go:
```diff
@@ -586,24 +581,21 @@ var _ = SIGDescribe("Services", func() {
-		// TODO: write a wrapper for ExpectNoErrorWithOffset()
-		framework.ExpectNoErrorWithOffset(0, err)
+		nodeIP, err := e2enode.PickIP(cs) // for later
+		framework.ExpectNoError(err)
```
could you add a simple annotation here as well, please?

alejandrox1 commented on test/e2e/network/service.go:
```diff
@@ -586,24 +581,21 @@ var _ = SIGDescribe("Services", func() {
-		jig.SanityCheckService(tcpService, v1.ServiceTypeClusterIP)
+		tcpJig := e2eservice.NewTestJig(cs, ns1, serviceName)
+		tcpService, err := tcpJig.CreateTCPService(nil)
+		framework.ExpectNoError(err)
```
could you add a simple annotation here as well, please?

alejandrox1 commented on test/e2e/network/service.go:
```diff
@@ -586,24 +581,21 @@ var _ = SIGDescribe("Services", func() {
-		jig.SanityCheckService(udpService, v1.ServiceTypeClusterIP)
+		udpJig := e2eservice.NewTestJig(cs, ns2, serviceName)
+		udpService, err := udpJig.CreateUDPService(nil)
+		framework.ExpectNoError(err)
```
could you add a simple annotation here as well, please?

alejandrox1 commented on test/e2e/network/service.go:
```diff
@@ -613,26 +605,28 @@ var _ = SIGDescribe("Services", func() {
 		ginkgo.By("creating a pod to be part of the TCP service " + serviceName)
-		jig.RunOrFail(ns1, nil)
+		_, err = tcpJig.Run(nil)
+		framework.ExpectNoError(err)
```
could you add a simple annotation here as well, please?

alejandrox1 commented on test/e2e/network/service.go:
```diff
@@ -613,26 +605,28 @@ var _ = SIGDescribe("Services", func() {
 		ginkgo.By("creating a pod to be part of the UDP service " + serviceName)
-		jig.RunOrFail(ns2, nil)
+		_, err = udpJig.Run(nil)
+		framework.ExpectNoError(err)
```
could you add a simple annotation here as well, please?

alejandrox1 commented on test/e2e/network/service.go:
```diff
@@ -613,26 +605,28 @@ var _ = SIGDescribe("Services", func() {
 			s.Spec.Type = v1.ServiceTypeNodePort
 		})
-		jig.SanityCheckService(tcpService, v1.ServiceTypeNodePort)
+		framework.ExpectNoError(err)
```
could you add a simple annotation here as well, please?

alejandrox1 commented on test/e2e/network/service.go:
```diff
@@ -613,26 +605,28 @@ var _ = SIGDescribe("Services", func() {
 			s.Spec.Type = v1.ServiceTypeNodePort
 		})
-		jig.SanityCheckService(udpService, v1.ServiceTypeNodePort)
+		framework.ExpectNoError(err)
```
could you add a simple annotation here as well, please?

alejandrox1 commented on test/e2e/network/service.go:
```diff
@@ -672,16 +666,18 @@ var _ = SIGDescribe("Services", func() {
 			s.Spec.LoadBalancerIP = requestedIP // will be "" if not applicable
 			s.Spec.Type = v1.ServiceTypeLoadBalancer
 		})
+		framework.ExpectNoError(err)
```
could you add a simple annotation here as well, please?

alejandrox1 commented on test/e2e/network/service.go:
```diff
@@ -672,16 +666,18 @@ var _ = SIGDescribe("Services", func() {
+			udpService, err = udpJig.UpdateService(func(s *v1.Service) {
 				s.Spec.Type = v1.ServiceTypeLoadBalancer
 			})
+			framework.ExpectNoError(err)
```
could you add a simple annotation here as well, please?

alejandrox1 commented on test/e2e/network/service.go:
```diff
@@ -690,8 +686,8 @@ var _ = SIGDescribe("Services", func() {
-		tcpService = jig.WaitForLoadBalancerOrFail(ns1, tcpService.Name, loadBalancerCreateTimeout)
-		jig.SanityCheckService(tcpService, v1.ServiceTypeLoadBalancer)
+		tcpService, err = tcpJig.WaitForLoadBalancer(loadBalancerCreateTimeout)
+		framework.ExpectNoError(err)
```
could you add a simple annotation here as well, please?

alejandrox1 commented on test/e2e/network/service.go:
```diff
@@ -723,8 +719,8 @@ var _ = SIGDescribe("Services", func() {
-			udpService = jig.WaitForLoadBalancerOrFail(ns2, udpService.Name, loadBalancerCreateTimeout)
-			jig.SanityCheckService(udpService, v1.ServiceTypeLoadBalancer)
+			udpService, err = udpJig.WaitForLoadBalancer(loadBalancerCreateTimeout)
+			framework.ExpectNoError(err)
```
could you add a simple annotation here as well, please?

alejandrox1 commented on test/e2e/network/service.go:
```diff
@@ -754,8 +750,8 @@ var _ = SIGDescribe("Services", func() {
-		tcpService = jig.ChangeServiceNodePortOrFail(ns1, tcpService.Name, tcpNodePort)
-		jig.SanityCheckService(tcpService, v1.ServiceTypeLoadBalancer)
+		tcpService, err = tcpJig.ChangeServiceNodePort(tcpNodePort)
+		framework.ExpectNoError(err)
```
could you add a simple annotation here as well, please?

alejandrox1 commented on test/e2e/network/service.go:
```diff
@@ -767,12 +763,8 @@ var _ = SIGDescribe("Services", func() {
-			jig.SanityCheckService(udpService, v1.ServiceTypeNodePort)
-		}
+		udpService, err = udpJig.ChangeServiceNodePort(udpNodePort)
+		framework.ExpectNoError(err)
```
could you add a simple annotation here as well, please?

alejandrox1 commented on test/e2e/network/service.go:
```diff
@@ -806,10 +798,10 @@ var _ = SIGDescribe("Services", func() {
 			s.Spec.Ports[0].Port++
 		})
-		jig.SanityCheckService(tcpService, v1.ServiceTypeLoadBalancer)
+		framework.ExpectNoError(err)
```
could you add a simple annotation here as well, please?

alejandrox1 commented on test/e2e/network/service.go:
```diff
@@ -1956,23 +1976,14 @@ var _ = SIGDescribe("Services", func() {
-		if err != nil {
-			framework.Failf("ginkgo.Failed to create service: %v", err)
-		}
+		_, err = jig.CreateTCPServiceWithPort(nil, int32(port))
```
How are labels being set here?

alejandrox1 reviewed (commented):
This looks awesome! Thank you @danwinship !

A couple comments:
1. could you please add annotations to describe the behavior one would expect in `framework.ExpectNoError()`s?
2. Seems likeyou replaced `CreateServiceWithServicePort()` throughout the place. One thing this function allowed to do was to set the labels for the service. This functionality seems missing, is that correct?

danwinship commented:
> Seems like you replaced `CreateServiceWithServicePort()` throughout the place. One thing this function allowed to do was to set the labels for the service. This functionality seems missing, is that correct?

`TestJig` already handled this. `CreateServiceWithServicePort()` just ignored the built-in label support and reimplemented it itself (another way that it didn't fit in with the rest of `TestJig`)

danwinship commented on test/e2e/framework/ingress/ingress_utils.go:
```diff
@@ -841,17 +841,18 @@ func (cont *NginxIngressController) Init() {
 	})
-	cont.lbSvc = serviceJig.WaitForLoadBalancerOrFail(cont.Ns, "nginx-ingress-lb", e2eservice.GetServiceLoadBalancerCreationTimeout(cont.Client))
-	serviceJig.SanityCheckService(cont.lbSvc, v1.ServiceTypeLoadBalancer)
+	framework.ExpectNoError(err)
```
The error itself already says all of that though (`"failed to create TCP service 'nginx-ingress-lb': ..."`)...

danwinship commented on test/e2e/framework/ingress/ingress_utils.go:
```diff
@@ -841,17 +841,18 @@ func (cont *NginxIngressController) Init() {
-	serviceJig.SanityCheckService(cont.lbSvc, v1.ServiceTypeLoadBalancer)
+	framework.ExpectNoError(err)
+	cont.lbSvc, err = serviceJig.WaitForLoadBalancer(e2eservice.GetServiceLoadBalancerCreationTimeout(cont.Client))
+	framework.ExpectNoError(err)
```
likewise, the actual error would be `"timed out waiting for service 'nginx-ingress-lb' to have a load balancer"`

danwinship commented:
I can add descriptions if you think it's necessary, but I don't see how

    failed to create a load balancer service for nginx ingress controller
    Expected error:
        <*errors.errorString | 0xc820412c10>: {
            s: "failed to create TCP service 'nginx-ingress-lb': ...",
        }
        timed out waiting for the condition
    not to have occurred

is especially better than

    Expected error:
        <*errors.errorString | 0xc820412c10>: {
            s: "failed to create TCP service 'nginx-ingress-lb': ...",
        }
        timed out waiting for the condition
    not to have occurred

I am aware of #34059 but it seems like making *everyone* pass a description to `ExpectNoError` just because `k8s.io/apimachinery/pkg/util/wait` returns terrible error messages is an overreaction.

alejandrox1 reviewed (approved):
Ah yeah, I think you have a point in both areas.
the errors are already informative given the way you refactored the sanity checks and all else.
As for the labels, I did just notice that those can be set from the beginning.
SO, again, thank you for working on this.
Now we just need one more approval to get this in

/lgtm
/milestone v1.17
/assign @oomichi @timothysc 
ptal

oomichi commented:
Thanks for doing this.

/approve

cblecker commented:
/lgtm cancel

This is failing tests legitimately.

oomichi commented:
Yes, right.
https://github.com/kubernetes/kubernetes/pull/80725 which uses NewTestJig() and other functions has been merged after proposing this PR.
So we need to update test/e2e/network/dual_stack.go to avoid the following errors:
```
/b/f/w/test/e2e/network/dual_stack.go:213:31: not enough arguments in call to service.NewTestJig
	have (kubernetes.Interface, string)
	want (kubernetes.Interface, string, string)
/b/f/w/test/e2e/network/dual_stack.go:232:24: too many arguments in call to jig.CreateServicePods
	have (kubernetes.Interface, string, number)
	want (int)
/b/f/w/test/e2e/network/dual_stack.go:254:31: not enough arguments in call to service.NewTestJig
	have (kubernetes.Interface, string)
	want (kubernetes.Interface, string, string)
/b/f/w/test/e2e/network/dual_stack.go:268:24: too many arguments in call to jig.CreateServicePods
	have (kubernetes.Interface, string, number)
	want (int)
/b/f/w/test/e2e/network/dual_stack.go:290:31: not enough arguments in call to service.NewTestJig
	have (kubernetes.Interface, string)
	want (kubernetes.Interface, string, string)
/b/f/w/test/e2e/network/dual_stack.go:304:24: too many arguments in call to jig.CreateServicePods
	have (kubernetes.Interface, string, number)
	want (int)
```
as https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/83549/pull-kubernetes-bazel-build/1184904234366144512/build-log.txt

danwinship commented:
omg those new tests are using `e2eservice.TestJig` and `e2eservice.TestFixture` at the same time?! Anyway, I'll fix that when I get rid of `TestFixture`.
diff after the rebase: https://github.com/danwinship/kubernetes/compare/18e3b2c1..f53735f6

danwinship commented:
Rebased, passed tests, got needs-rebase'd again, re-rebased, passing tests again, needs /lgtm

oomichi commented:
Thanks for updating

/lgtm
